List GRPC services defined by protos

jds106 opened the pull request:
As discussed, I have added a new `command` parameter. This currently only takes a single value: `list_services`. It is expected that `polyglot` will move over to a "command"-based invocation in the future, and then depend on the Flagz library. 

The `list_services` command takes 3 optional arguments:

```
--service_filter <service name substring>
--method_filter <method name substring>
--with_message <true|false>
```

The two filter arguments run a case-insensitive "contains" filter. If `with-message` is specified, then a simple message is rendered to make it easier for a user to construct a message to send to an end-point.

Commit:
Added functionality to list service meta data

Commit:
Added functionality to list service meta data

Commit:
Fixed legacy command method

Commit:
Fixed legacy command method

Commit:
Fixed legacy command method

Commit:
Fixed file location of protos

dinowernli commented on src/main/java/polyglot/Main.java:
```diff
@@ -64,12 +64,31 @@ public static void main(String[] args) {
+    
+    } else {
+      switch (arguments.command().get()) {
+        case "list_services":
```
Can we make this a private static final constant?

jds106 replied:
Done

dinowernli commented on src/main/java/polyglot/Main.java:
```diff
@@ -64,12 +64,31 @@ public static void main(String[] args) {
+      runLegacyCommands(fileDescriptorSet, config, arguments);
+    
+    } else {
+      switch (arguments.command().get()) {
```
Could we also have a "call" command for the previous calling behavior? Then, we could model "call" as simply being the default.

jds106 replied:
Done

dinowernli commented on src/main/java/polyglot/config/CommandLineArgs.java:
```diff
@@ -50,6 +50,29 @@
+  // *************************************************************************
+  // * Initial step towards the migration to "polyglot <command> [flagz...]" *
+  // *************************************************************************
+  @Option(name = "--command", metaVar = "<e.g. list_services>")	// Expand to support many commands
```
Could we have the legal values in the metaVar (like for --with_message below)? Something like "list_services|call"

jds106 replied:
Done

dinowernli commented on src/main/java/polyglot/config/CommandLineArgs.java:
```diff
@@ -50,6 +50,29 @@
+  
+  // TODO: Move to a "list_services"-specific flag container
+  @Option(name = "--method_filter", metaVar = "method_name", 
+		  usage="Filters methods on a service to those containing this string")
```
nit: indent. The one below is correct

jds106 replied:
I think that was an issue with tabs. What layout is standard? I have currently changed it to this:

```
  // TODO: Move to a "list_services"-specific flag container
  @Option(name = "--method_filter", metaVar = "method_name", 
          usage="Filters methods on a service to those containing this string")
  private String methodFilterArg;
```

dinowernli commented on src/main/java/polyglot/config/CommandLineArgs.java:
```diff
@@ -50,6 +50,29 @@
+  
+  // TODO: Move to a "list_services"-specific flag container
+  @Option(name = "--service_filter", metaVar = "service_name", 
+		  usage="Filters services to those containing this string")
```
Please give an example of the format in usage.

jds106 replied:
Done

dinowernli commented on src/main/java/polyglot/config/CommandLineArgs.java:
```diff
@@ -89,6 +112,11 @@ private CommandLineArgs() {
 
   private void initialize() {
+	// Short-circuit the preconditions checks if the "command" flag is present (as flag validation is command-specific) 
+	if (command().isPresent()) {
```
Looks like this might be dangerous if we add more commands. Could we have a command-specific validation?

dinowernli commented on src/main/java/polyglot/config/CommandLineArgs.java:
```diff
@@ -136,6 +164,46 @@ public ProtoMethodName grpcMethodName() {
+   * 	list_services [--service_filter XXX] [--method_filter YYY] 
+   */
+  public Optional<String> command() {
+	  if (commandArg == null) {
```
return Optional.ofNullable(commandArg)?

also, indent :) [and below]

jds106 replied:
Done

dinowernli commented on src/main/java/polyglot/ServiceList.java:
```diff
@@ -0,0 +1,108 @@
+    ServiceResolver serviceResolver = ServiceResolver.fromFileDescriptorSet(fileDescriptorSet);
+
+    // Add white-space before the rendered output
+    System.out.println();
```
So far, we have been using the logger for all output. That approach needs some re-thinking, but we should probably avoid "System.out.println" for now for consistency (unless you think this is unreasonable for your use case). Thoughts?

jds106 replied:
This isn't logging - it's functional output to stdout that you might subsequently want to pipe to another tool (e.g. grep).

My personal preference is to have logging on stderr (which it currently is), and output that is to be used by the user (either directly or indirectly) is written as "clean text" (i.e. with no logging meta data such as time etc.) to stdout. Happy to discuss - and ultimately you have the casting vote.

dinowernli commented on src/main/java/polyglot/Main.java:
```diff
@@ -64,12 +64,31 @@ public static void main(String[] args) {
+  }
+  
+  /** The "legacy commands" will be deprecated when we move over to a "polyglot [command]" style invocation */
+  private static void runLegacyCommands(FileDescriptorSet fileDescriptorSet, final Configuration config, final CommandLineArgs arguments) {
```
Long line (here an elsewhere). We're sticking to <= 100 :)

jds106 replied:
Fixed all long lines.

dinowernli commented on src/main/java/polyglot/ServiceList.java:
```diff
@@ -0,0 +1,108 @@
+import polyglot.protobuf.ServiceResolver;
+
+/** Simple utility to list the services, methods and message definitions for the known GRPC end-points */
+public class ServiceList {
```
A very simple unit test would be nice, to make sure future changes don't break this.

jds106 replied:
Added

dinowernli commented on src/main/java/polyglot/ServiceList.java:
```diff
@@ -0,0 +1,108 @@
+package polyglot;
```
Maybe move this to the same package as the ServiceResolver?

jds106 replied:
Done

dinowernli commented:
@jds106 looking really nice. A few comments, nothing major though.

Commit:
Refactor based on PR

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Added unit test, formatted line lengths

jds106 commented:
This PR has become rather large. The summary is:
- Extracted the logic to make a GRPC call into a `polyglot.command.ServiceCall`
- Moved `ServiceList` into the same package (`polyglot.command`)
- Updated the BUILD files in `polyglot` and `polyglot.command` with the correct dependencies
- Moved `ServiceCall`-specific command-line logic out of `CommmandLineArgs` into `ServiceCall`
- Added new unit test to test `ServiceCall` output: `polyglot.command.ServiceListTest`

It may be easier to look at these files in their entirety as the diffs below are rather messy. More than happy to simply pair-up to make any changes you'd like.

dinowernli commented on src/main/java/polyglot/Main.java:
```diff
@@ -64,54 +40,38 @@ public static void main(String[] args) {
+    switch (command) {
+      case CommandLineArgs.LIST_SERVICES_COMMAND:
+        ServiceList.listServices(
+          fileDescriptorSet, config.getProtoConfig().getProtoDiscoveryRoot(), 
```
nit: indent by two more

jds106 replied:
Done

dinowernli commented on src/main/java/polyglot/command/BUILD:
```diff
@@ -0,0 +1,18 @@
+  name = "command",
+  srcs = glob(['*.java']),
+  deps = [
+   "//src/main/proto:config_proto",
```
nit: indent by 1 more

jds106 replied:
Done

dinowernli commented on src/main/java/polyglot/protobuf/ServiceResolver.java:
```diff
@@ -43,6 +44,15 @@ public static ServiceResolver fromFileDescriptorSet(FileDescriptorSet descriptor
+  
+  /** Lists all of the services found in the file descriptors */
+  public Iterable<ServiceDescriptor> listServices() {
+      ArrayList<ServiceDescriptor> serviceDescriptors = new ArrayList<ServiceDescriptor>(); 
```
indent (2 spaces, not 4)

jds106 replied:
DOH! Fixed :-)

dinowernli commented on src/test/java/polyglot/command/ServiceListTest.java:
```diff
@@ -0,0 +1,115 @@
+      validateOutput(baos.toString(), 
+          expectedServiceName, new String[]{"TestMethod", "TestMethodStream", "TestMethodBidi"});
+      
+      baos.reset();
```
Is it worth splitting these into individual tests?

jds106 replied:
Done

dinowernli commented:
@jds106 just tiny nits left, LGTM after those

Commit:
Split test case, fixed up formatting

dinowernli commented:
Thanks for your contribution!
